olp-cpp-sdk-authentication/src/AuthenticationClient.cpp

```diff
@@ -450,206 +451,124 @@ AuthenticationClient::Impl::~Impl() { pending_requests_->CancelAllAndWait(); }
-    request.WithSettings(std::move(network_settings));
-
-    std::shared_ptr<std::stringstream> payload =
-        std::make_shared<std::stringstream>();
-    request.WithBody(GenerateClientBody(properties));
-
-    auto network_callback = [callback, payload, credentials, cache](
-                                const http::NetworkResponse& network_response) {
-      auto response_status = network_response.GetStatus();
-      auto error_msg = network_response.GetError();
-
-      // Network not available, use cached token if available
-      if (response_status < 0) {
-        // Request cancelled, return
-        if (response_status ==
-            static_cast<int>(olp::http::ErrorCode::CANCELLED_ERROR)) {
-          callback({{response_status, error_msg}});
-          return;
-        }
 
-        auto cached_response_found =
-            cache->locked([credentials, callback](
-                              utils::LruCache<std::string, SignInResult>& c) {
-              auto it = c.Find(credentials.GetKey());
-              if (it != c.end()) {
-                SignInClientResponse response(it->value());
-                callback(response);
-                return true;
-              }
-              return false;
-            });
-
-        if (!cached_response_found) {
-          // Return an error response
-          SignInClientResponse error_response(
-              AuthenticationError(response_status, error_msg));
-          callback(error_response);
-        }
+    const auto url = settings_.token_endpoint_url + kOauthEndpoint;
+
+    client::OlpClient::ParametersType headers = {
+        {http::kAuthorizationHeader,
+         GenerateHeader(credentials, url, timestamp)}};
+
+    auto auth_response = client.CallApi(
+        kOauthEndpoint, "POST", {}, std::move(headers), {},
+        GenerateClientBody(properties), kApplicationJson, context);
 
-        return;
+    const auto status = auth_response.status;
+
+    if (status < 0) {
+      auto error_message = auth_response.response.str();
+
+      if (status == static_cast<int>(olp::http::ErrorCode::CANCELLED_ERROR)) {
+        return {{status, error_message}};
       }
 
-      auto document = std::make_shared<rapidjson::Document>();
-      rapidjson::IStreamWrapper stream(*payload);
-      document->ParseStream(stream);
-
-      std::shared_ptr<SignInResultImpl> resp_impl =
-          std::make_shared<SignInResultImpl>(response_status, error_msg,
-                                             document);
-      SignInResult response(resp_impl);
-
-      if (response_status == http::HttpStatusCode::OK) {
-        // Cache the response
-        cache->locked([credentials, &response](
-                          utils::LruCache<std::string, SignInResult>& c) {
-          return c.InsertOrAssign(credentials.GetKey(), response);
-        });
+      auto cached_response = client_token_cache_->locked(
+          [credentials](utils::LruCache<std::string, SignInResult>& c)
```

Comment:
```suggestion
          [&](utils::LruCache<std::string, SignInResult>& c)
```
Sync call, we do not need to copy it.